### ISSUE TITLE
Faster array access

### ISSUE THREAD

**Issue by Serggio** [X-Sharp/XSharpPublic#400]
Currently, the index property in __Array class is defined like this:
```
PUBLIC PROPERTY SELF[index PARAMS INT[]] AS USUAL
```
which calls the __GetElement method defined this way:
```
PUBLIC METHOD __GetElement(index PARAMS INT[]) AS USUAL
```

Both of them use PARAMS. And that makes quite a big overhead when accessing one or two-dimensional Arrays, which are the most frequently used ones.
Behind the scenes there is a List<INT>[] generated that will be passed to that index-property as a PARAMS list, it calls "newarr" IL-command, which in turn calls malloc, and that is just to send a single index number to the getter!
The idea is to add some overloaded versions of an index-property and the __GetElement method, and make them accept a single index. And maybe some more overloads for two and three indexes. These overloads should NOT call the PARAMS versions of themselves (for that would devalue the whole idea) but implement access to the _internalList on their own. Thus, the compiler would link one or two-dimensional requests to these faster versions. And there still should be the universal PARAMS version for super-multi-dimensional calls.

I've made several tests: made a class similar to __Array - it has the same internal List<USUAL>, but gave it a single-argument index property (along-side with the PARAMS-version):
```
PUBLIC PROPERTY SELF[dwIndex AS DWORD] AS USUAL
```
I've made some substantial tests with it, calculating sum and average of elements (in order optimizer not to throw off needless code), and compared these results with the same algorithm but with ARRAY.
The result was **25-30% faster**. It is quite significant. It was a release build with optimizer on, etc.
Please consider this proposal, because adding these one or two overloaded methods is both simple and effective, and it will dramatically increase the speed of Arrays in the majority of use cases.

**Comment by RobertvanderHulst** [comment]
Serggio,
I certainly want t look at this.
Can you share your example (simply drag a ZIP onto the message pane) to include it as an atttachment ?

**Comment by Serggio** [comment]
Here you are Robert!
As a bonus I've made a third test with ARRAY OF INT - it seems to be... 40 times faster than the USUAL version! :) That is super cool but at the same time it testifies that there is a room to optimize USUALs either ;)

[ArrayTest.zip](https://github.com/X-Sharp/XSharpPublic/files/4671765/ArrayTest.zip)

**Comment by RobertvanderHulst** [comment]
Serggio,
I ran your test and I see almost no difference between the Fast Array and the normal Array.
I do see the difference between the Array and the ARRAY of INT.
And that is not a surprise: we always tell our customers to change USUALs to typed variables for performance and early error detection.

When I change the code to use AS ARRAY OF USUAL instead of the normal ARRAY type then that type is faster than your "FastArray" Class. The difference with the normal array is that indeed it uses the indexer on the __ArrayBase class with a single Int argument .

The reason why we added the multi dimensional index is to allow the syntax 
`aArray[1,2]`
Without that multi dimensional indexer this has to be written as  `aArray[1][2]` and requires 2 indexer calls. 

But it should not hurt to add a "simple" indexer as well, I think.
